### PR TITLE
Clang format

### PR DESCRIPTION
**Pull request by dtebbs**
- Clang format config and scripts.
- Most files are unedited after formatting, except where manual edits were required. Any other files can be addressed as they are touched / seen in later PRs.

NOTE: there seems to be some difference between clang-format version `11.0.0-2~ubuntu20.04.1` and `11.0.0`, related to lines of the form `#else // if DEBUG` (11.0.0.2 wants to put 2 spaces before the comment, which conflicts with 11.0.0).

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (AntoineRondelet) [inline .github/workflows/ci.yml]: Why not use the same code snippet as the one in Zeth: https://github.com/clearmatics/zeth/blob/develop/scripts/ci#L28-L35?

`[ "0" == cat format_errs | wc -l ]` -> `[ cat format_errs | wc -l  ==   "0" ]`

File: depends/.clang-format
Comment (AntoineRondelet) [inline depends/.clang-format]: You may want to add a final empty line here. (see: https://thoughtbot.com/blog/no-newline-at-end-of-file)

File: libsnark/common/data_structures/accumulation_vector.hpp
Comment (AntoineRondelet) [inline libsnark/common/data_structures/accumulation_vector.hpp:50]: Hum.. Have you done some manual tweaks here? Seems like the formatting isn't fully consistent here (see `{}` vs `{ \n }`)

File: libsnark/common/data_structures/merkle_tree.tcc
Comment (AntoineRondelet) [inline libsnark/common/data_structures/merkle_tree.tcc]: Let's try to avoid comments at the end of lines as it messes up with the format a little bit. It's better to comment above to avoid unnecessary line breaks.

File: libsnark/common/data_structures/sparse_vector.hpp
Comment (AntoineRondelet) [inline libsnark/common/data_structures/sparse_vector.hpp]: Let's move some of these comments up to avoid line breaks

File: libsnark/common/data_structures/sparse_vector.tcc
Comment (AntoineRondelet) [inline libsnark/common/data_structures/sparse_vector.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/common/data_structures/sparse_vector.tcc
Comment (AntoineRondelet) [inline libsnark/common/data_structures/sparse_vector.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/common/default_types/tinyram_zksnark_pp.cpp
Comment (AntoineRondelet) [inline libsnark/common/default_types/tinyram_zksnark_pp.cpp]: Add final line (see earlier comment)

File: libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp
Comment (AntoineRondelet) [inline libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp]: Let's move this comment up to avoid line breaks

File: libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp
Comment (AntoineRondelet) [inline libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp]: Use `//` instead of `/* */` syntax here to avoid unnecessary line break

File: libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp
Comment (AntoineRondelet) [inline libsnark/common/routing_algorithms/as_waksman_routing_algorithm.cpp]: Let's move this comment up to avoid line breaks

File: libsnark/common/routing_algorithms/benes_routing_algorithm.cpp
Comment (AntoineRondelet) [inline libsnark/common/routing_algorithms/benes_routing_algorithm.cpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/hashes/knapsack/knapsack_gadget.hpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/hashes/knapsack/knapsack_gadget.hpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/hashes/knapsack/knapsack_gadget.hpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/hashes/knapsack/knapsack_gadget.hpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_update_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_update_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_update_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_update_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/pairing/weierstrass_precomputation.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/routing/as_waksman_routing_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/routing/as_waksman_routing_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/routing/as_waksman_routing_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/routing/as_waksman_routing_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/set_commitment/set_commitment_gadget.tcc]: Let's move this comment up to avoid line breaks (see line above too)

File: libsnark/gadgetlib1/gadgets/verifiers/r1cs_ppzksnark_verifier_gadget.hpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/verifiers/r1cs_ppzksnark_verifier_gadget.hpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/gadgets/verifiers/r1cs_ppzksnark_verifier_gadget.tcc
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/verifiers/r1cs_ppzksnark_verifier_gadget.tcc]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/protoboard.hpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/protoboard.hpp]: Let's move this comment up to avoid line breaks

File: libsnark/gadgetlib1/protoboard.hpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/protoboard.hpp]: Let's move this comment up to avoid line breaks

File: libsnark/reductions/r1cs_to_qap/r1cs_to_qap.tcc
Comment (AntoineRondelet) [inline libsnark/reductions/r1cs_to_qap/r1cs_to_qap.tcc]: Let's move this comment up to avoid line breaks

**Review by AntoineRondelet — COMMENTED** [review]
Couple of issues with inline comments. I flagged some here. Please don't force push the branch @dtebbs to keep it easy to review till the PR is ready to be merged.

File: libsnark/common/data_structures/merkle_tree.tcc
Comment (dtebbs) [reply libsnark/common/data_structures/merkle_tree.tcc]: Yes.  I generally don't add end-of-line comments. As in the description, for now I intended to fix only those that I came across so there are several of these cases still in the code.

As well as those you've commented, I've tried to do a few greps to catch the remaining cases, although there may be some still to address.  The greps are slightly non-trivial since we have a few valid cases such as `} // namespace libsnark` to take care of.

File: libsnark/common/data_structures/accumulation_vector.hpp
Comment (dtebbs) [reply libsnark/common/data_structures/accumulation_vector.hpp:50]: No,  I did not change anything here (in general clang-format does allow many tweaks of this kind).  The inconsistency seems to be due to the unnecessary semi-colon, which I've removed.

File: depends/.clang-format
Comment (dtebbs) [reply depends/.clang-format]: Added.  Thanks.

File: .github/workflows/ci.yml
Comment (dtebbs) [reply .github/workflows/ci.yml]: Switched.

**Comment by dtebbs** [comment]
@AntoineRondelet These should all be addressed now.  Thanks.
If you can approve, once you are happy, I'll rebase to fix the conflicts and do the merge from my end (one the dependent PR is merged).  Thanks.

File: libsnark/common/data_structures/accumulation_vector.hpp
Comment (AntoineRondelet) [reply libsnark/common/data_structures/accumulation_vector.hpp:50]: >  The inconsistency seems to be due to the unnecessary semi-colon, which I've removed.

Ah! Thanks, I see

Comment (AntoineRondelet) [reply libsnark/common/data_structures/accumulation_vector.hpp:50]: Shall the `;` be removed on the line below too then? Or maybe, we should just add the `;` to keep the `{};` format and avoid the `{ \n }` pattern. What do you think?

File: libsnark/common/data_structures/merkle_tree.tcc
Comment (AntoineRondelet) [reply libsnark/common/data_structures/merkle_tree.tcc]: Yes I know lots of code changes here are related to "old" code, much of which we didn't write. No problem if there are a few instances of inline comments causing some minor formatting issues, we'll adjust as we go and as we touch the files concerned. I wanted to do a first pass in this PR, but if some of them escape our attention this is not a big deal for now.

File: libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp
Comment (AntoineRondelet) [inline libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp:42]: Was this space forced by clang-format?

File: libsnark/common/data_structures/accumulation_vector.hpp
Comment (dtebbs) [reply libsnark/common/data_structures/accumulation_vector.hpp:50]: Thanks.  I've removed the semi-colon.

I'd actually prefer it if the interface definition and the implementations were not mixed together in this way, i.e. all function implementations (including inlined) given outside of the class.  I tend to do this for new code, but it would be a huge change to apply this everywhere.

**Comment by AntoineRondelet** [comment]
> @AntoineRondelet These should all be addressed now. Thanks. If you can approve, once you are happy, I'll rebase to fix the conflicts and do the merge from my end (one the dependent PR is merged). Thanks.

Thanks for fixing these @dtebbs, there are just a few comments outstanding and we'll be able to roll afterwards.

File: libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp
Comment (dtebbs) [reply libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp:42]: Yes.  Comments seem to always be indented according to the current context.

File: libsnark/common/data_structures/accumulation_vector.hpp
Comment (AntoineRondelet) [reply libsnark/common/data_structures/accumulation_vector.hpp:50]: > I'd actually prefer it if the interface definition and the implementations were not mixed together in this way

Yeah I align with you on that.

> I tend to do this for new code, but it would be a huge change to apply this everywhere.

For sure, not one for this PR :)

>  I've removed the semi-colon.

Thanks

File: libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp
Comment (AntoineRondelet) [reply libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.cpp:42]: Got it, that kind of make sense actually, just doesn't work super well with these preprocessor directives, but no big deal.

**Review by AntoineRondelet — APPROVED** [review]
LGTM, I'll let you rebase @dtebbs and I'll merge then, after the CI checks pass.

**Pull request merged**
